crda/remotematcher.go

```diff
@@ -0,0 +1,258 @@
+//
+// If not passed to NewMatcher, a defaultRequestConcurrency will be used.
+func WithRequestConcurrency(requestConcurrency int) Option {
+	return func(m *Matcher) error {
+		m.requestConcurrency = requestConcurrency
+		return nil
+	}
+}
+
+// Name implements driver.Matcher.
+func (*Matcher) Name() string { return "crda" }
+
+// Filter implements driver.Matcher.
+func (*Matcher) Filter(record *claircore.IndexRecord) bool {
+	return record.Package.NormalizedVersion.Kind == "pep440"
+}
+
+// Query implements driver.Matcher.
+func (*Matcher) Query() []driver.MatchConstraint {
+	panic("unreachable")
+}
+
+// Vulnerable implements driver.Matcher.
+func (*Matcher) Vulnerable(ctx context.Context, record *claircore.IndexRecord, vuln *claircore.Vulnerability) (bool, error) {
+	// RemoteMatcher can match Package and Vulnerability.
+	panic("unreachable")
+}
+
+// QueryRemoteMatcher implements driver.RemoteMatcher.
+func (m *Matcher) QueryRemoteMatcher(ctx context.Context, records []*claircore.IndexRecord) (map[string][]*claircore.Vulnerability, error) {
+	log := zerolog.Ctx(ctx).With().
+		Str("component", "crda/remotematcher/QueryRemoteMatcher").
+		Logger()
+	ctx = log.WithContext(ctx)
+	log.Debug().
+		Int("records", len(records)).
+		Msg("packages")
+
+	ctrlC, errorC := m.fetchVulnerabilities(ctx, records)
+	results := make(map[string][]*claircore.Vulnerability)
+	for r := range ctrlC {
+		for _, vuln := range r {
+			results[vuln.Package.ID] = append(results[vuln.Package.ID], vuln)
+			log.Debug().
+				Str("package", vuln.Package.Name).
+				Str("version", vuln.Package.Version).
+				Str("id", vuln.ID).
+				Msg("vulns")
+		}
+	}
+	select {
+	case err, ok := <-errorC:
+		// Don't propagate error, log and move on.
+		if ok {
+			log.Error().Err(err).Msg("access to component analyses failed")
+		}
+	default:
+	}
+	log.Debug().
+		Int("vulnerabilities", len(results)).
+		Msg("query")
+	return results, nil
+}
+
+func (m *Matcher) fetchVulnerabilities(ctx context.Context, records []*claircore.IndexRecord) (chan []*claircore.Vulnerability, chan error) {
+	inC := make(chan *claircore.IndexRecord, m.requestConcurrency)
+	ctrlC := make(chan []*claircore.Vulnerability, m.requestConcurrency)
+	errorC := make(chan error, 1)
+	go func() {
+		defer close(ctrlC)
+		defer close(errorC)
+		var g errgroup.Group
+		for _, record := range records {
+			g.Go(func() error {
+				vulns, err := m.componentAnalyses(ctx, <-inC)
+				if err != nil {
+					return err
+				}
+				ctrlC <- vulns
+				return nil
+			})
+			inC <- record
+		}
+		close(inC)
+		if err := g.Wait(); err != nil {
+			errorC <- err
+		}
+	}()
+	return ctrlC, errorC
+}
+
+func (m *Matcher) componentAnalyses(ctx context.Context, record *claircore.IndexRecord) ([]*claircore.Vulnerability, error) {
+	reqUrl := url.URL{
+		Scheme:   m.url.Scheme,
+		Host:     m.url.Host,
+		Path:     fmt.Sprintf("/api/v2/component-analyses/pypi/%s/%s", record.Package.Name, record.Package.Version),
```

Comment:
probably want this string path as a constant declare higher in the file.

Reply:
Ack!